core/shared/src/main/scala/zio/stm/TMap.scala

```diff
@@ -0,0 +1,211 @@
+      if (toRemove.isEmpty) STM.succeed(toRetain) else tSize.update(_ - toRemove.size).as(toRetain)
+    }
+
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      _       <- buckets.updateM(idx, removeMatching)
+    } yield ()
+  }
+
+  /**
+   * Atomically folds using pure function.
+   */
+  final def fold[A](zero: A)(op: (A, (K, V)) => A): STM[Nothing, A] =
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.fold(zero)((acc, bucket) => bucket.foldLeft(acc)(op))
+    } yield res
+
+  /**
+   * Atomically folds using effectful function.
+   */
+  final def foldM[A, E](zero: A)(op: (A, (K, V)) => STM[E, A]): STM[E, A] = {
+    def loopM(acc: STM[E, A], remaining: List[(K, V)]): STM[E, A] =
+      remaining match {
+        case Nil          => acc
+        case head :: tail => loopM(acc.flatMap(op(_, head)), tail)
+      }
+
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.foldM(zero)((acc, bucket) => loopM(STM.succeed(acc), bucket))
+    } yield res
+  }
+
+  /**
+   * Atomically performs side-effect for each binding present in map.
+   */
+  final def foreach[E](f: ((K, V)) => STM[E, Unit]): STM[E, Unit] =
+    self.foldM(())((_, kv) => f(kv))
+
+  /**
+   * Retrieves value associated with given key.
+   */
+  final def get[E](k: K): STM[E, Option[V]] =
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      bucket  <- buckets(idx)
+    } yield bucket.find(_._1 == k).map(_._2)
+
+  /**
+   * Retrieves value associated with given key or default value, in case the
+   * key isn't present.
+   */
+  final def getOrElse[E](k: K, default: => V): STM[E, V] =
+    get(k).map(_.getOrElse(default))
+
+  /**
+   * Stores new binding into the map.
+   */
+  final def put[E](k: K, v: V): STM[E, Unit] = {
+    def upsert(bucket: List[(K, V)]): STM[Nothing, List[(K, V)]] = {
+      val exists = bucket.find(_._1 == k).isDefined
+
+      if (exists)
+        STM.succeed(bucket.map(kv => if (kv._1 == k) (k, v) else kv))
+      else
+        tSize.update(_ + 1).as((k, v) :: bucket)
+    }
+
+    def resize(newCapacity: Int): STM[Nothing, Unit] =
+      for {
+        data       <- self.fold(List.empty[(K, V)])((acc, kv) => kv :: acc)
+        tmap       <- TMap.allocate(newCapacity, data)
+        newBuckets <- tmap.tBuckets.get
+        _          <- self.tBuckets.set(newBuckets)
+        _          <- self.tCapacity.set(newCapacity)
+      } yield ()
+
+    for {
+      buckets     <- tBuckets.get
+      idx         <- indexOf(k)
+      _           <- buckets.updateM(idx, upsert)
+      size        <- tSize.get
+      capacity    <- tCapacity.get
+      needsResize = capacity * TMap.LoadFactor < size
+      _           <- if (needsResize) resize(capacity * 2) else STM.unit
+    } yield ()
+  }
+
+  /**
+   * Atomically updates all bindings using pure function.
+   */
+  final def transform(f: ((K, V)) => (K, V)): STM[Nothing, Unit] =
```

Comment:
Does it make sense to also include mapValues?

Reply:
I'm undecided about this one, clearly it's a low-hanging fruit:

```scala
final def mapValues(f: V => V): STM[Nothing, Unit] =
  transform(kv => kv._1 -> f(kv._2))
```

The reason I'm hesitating to put it is the API "consistency" as we've removed `map` / `mapM` already.

Reply:
Should be called transformValues, right?

Reply:
Introduced `transformValues` and `transformValuesM`.